File: DatabaseConfig.cs
/**
 * MaraBet AI - Database Configuration (C# / .NET)
 * Obtém credenciais do RDS PostgreSQL via AWS Secrets Manager
 */

using System;
using System.Threading.Tasks;
using Amazon;
using Amazon.SecretsManager;
using Amazon.SecretsManager.Model;
using Npgsql;
using Newtonsoft.Json;
using System.Collections.Generic;

namespace MaraBet.Config
{
    /// <summary>
    /// Gerenciador de configuração do banco de dados RDS
    /// </summary>
    public class DatabaseConfig
    {
        private const string SecretName = "rds!db-3758a324-12a2-4675-b5ff-b92acdf38483";
        private const string RegionName = "eu-west-1";
        private const string DatabaseName = "marabet_production";
        
        private Dictionary<string, string> _credentials;
        private static DatabaseConfig _instance;
        private static readonly object _lock = new object();
        
        /// <summary>
        /// Singleton instance
        /// </summary>
        public static DatabaseConfig Instance
        {
            get
            {
                if (_instance == null)
                {
                    lock (_lock)
                    {
                        if (_instance == null)
                        {
                            _instance = new DatabaseConfig();
                        }
                    }
                }
                return _instance;
            }
        }
        
        private DatabaseConfig()
        {
            _credentials = null;
        }
        
        /// <summary>
        /// Obtém credenciais do AWS Secrets Manager
        /// </summary>
        public async Task<Dictionary<string, string>> GetSecretAsync()
        {
            var region = RegionEndpoint.GetBySystemName(RegionName);
            var client = new AmazonSecretsManagerClient(region);
            
            var request = new GetSecretValueRequest
            {
                SecretId = SecretName
            };
            
            GetSecretValueResponse response;
            
            try
            {
                response = await client.GetSecretValueAsync(request);
            }
            catch (ResourceNotFoundException)
            {
                throw new Exception($"Secret {SecretName} não encontrado");
            }
            catch (InvalidRequestException e)
            {
                throw new Exception($"Requisição inválida: {e.Message}");
            }
            catch (InvalidParameterException e)
            {
                throw new Exception($"Parâmetro inválido: {e.Message}");
            }
            catch (DecryptionFailureException e)
            {
                throw new Exception($"Falha ao descriptografar: {e.Message}");
            }
            catch (InternalServiceErrorException e)
            {
                throw new Exception($"Erro interno AWS: {e.Message}");
            }
            
            var secretString = response.SecretString;
            
            // Parse JSON
            var secret = JsonConvert.DeserializeObject<Dictionary<string, string>>(secretString);
            
            return secret;
        }
        
        /// <summary>
        /// Obtém credenciais (com cache)
        /// </summary>
        public async Task<Dictionary<string, string>> GetCredentialsAsync()
        {
            if (_credentials == null)
            {
                _credentials = await GetSecretAsync();
            }
            return _credentials;
        }
        
        /// <summary>
        /// Gera connection string para PostgreSQL
        /// </summary>
        public async Task<string> GetConnectionStringAsync(string database = null)
        {
            var creds = await GetCredentialsAsync();
            var dbName = database ?? DatabaseName;
            
            return $"Host={creds["host"]};" +
                   $"Port={creds["port"]};" +
                   $"Database={dbName};" +
                   $"Username={creds["username"]};" +
                   $"Password={creds["password"]};" +
                   $"SSL Mode=Require;" +
                   $"Trust Server Certificate=true";
        }
        
        /// <summary>
        /// Obtém conexão Npgsql ao banco de dados
        /// </summary>
        public async Task<NpgsqlConnection> GetConnectionAsync(string database = null)
        {
            var connectionString = await GetConnectionStringAsync(database);
            var connection = new NpgsqlConnection(connectionString);
            await connection.OpenAsync();
            return connection;
        }
        
        /// <summary>
        /// Testa conexão com o banco
        /// </summary>
        public async Task<bool> TestConnectionAsync()
        {
            try
            {
                using (var conn = await GetConnectionAsync())
                {
                    using (var cmd = new NpgsqlCommand("SELECT version()", conn))
                    {
                        var version = await cmd.ExecuteScalarAsync();
                        Console.WriteLine("✅ Conexão bem-sucedida!");
                        Console.WriteLine($"   PostgreSQL: {version.ToString().Substring(0, 50)}...");
                    }
                }
                return true;
            }
            catch (Exception e)
            {
                Console.WriteLine($"❌ Erro na conexão: {e.Message}");
                return false;
            }
        }
        
        /// <summary>
        /// Health check do banco de dados
        /// </summary>
        public async Task<Dictionary<string, object>> HealthCheckAsync()
        {
            try
            {
                using (var conn = await GetConnectionAsync())
                {
                    // Versão
                    using (var cmd = new NpgsqlCommand("SELECT version()", conn))
                    {
                        var version = await cmd.ExecuteScalarAsync();
                        
                        // Conexões ativas
                        cmd.CommandText = "SELECT count(*) FROM pg_stat_activity";
                        var connections = await cmd.ExecuteScalarAsync();
                        
                        // Tamanho do database
                        cmd.CommandText = $"SELECT pg_size_pretty(pg_database_size('{DatabaseName}'))";
                        var size = await cmd.ExecuteScalarAsync();
                        
                        return new Dictionary<string, object>
                        {
                            ["status"] = "healthy",
                            ["version"] = version,
                            ["activeConnections"] = connections,
                            ["databaseSize"] = size,
                            ["timestamp"] = DateTime.UtcNow
                        };
                    }
                }
            }
            catch (Exception e)
            {
                return new Dictionary<string, object>
                {
                    ["status"] = "unhealthy",
                    ["error"] = e.Message,
                    ["timestamp"] = DateTime.UtcNow
                };
            }
        }
        
        /// <summary>
        /// Imprime informações do banco
        /// </summary>
        public async Task PrintInfoAsync()
        {
            var creds = await GetCredentialsAsync();
            
            Console.WriteLine(new string('=', 70));
            Console.WriteLine("🗄️  MARABET AI - RDS POSTGRESQL");
            Console.WriteLine(new string('=', 70));
            Console.WriteLine();
            Console.WriteLine($"Host:         {creds["host"]}");
            Console.WriteLine($"Port:         {creds["port"]}");
            Console.WriteLine($"Username:     {creds["username"]}");
            Console.WriteLine($"Password:     {new string('*', creds["password"].Length)}");
            Console.WriteLine($"Engine:       {creds["engine"]}");
            Console.WriteLine($"Database:     {DatabaseName}");
            Console.WriteLine();
            Console.WriteLine("Connection String:");
            Console.WriteLine(new string('-', 70));
            Console.WriteLine(await GetConnectionStringAsync());
            Console.WriteLine();
            Console.WriteLine(new string('=', 70));
        }
    }
    
    /// <summary>
    /// Classe principal para teste
    /// </summary>
    class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("🚀 MaraBet AI - Database Configuration\n");
            
            try
            {
                var config = DatabaseConfig.Instance;
                
                // Mostrar informações
                await config.PrintInfoAsync();
                
                // Testar conexão
                Console.WriteLine("\n🔌 Testando conexão...");
                Console.WriteLine(new string('-', 70));
                await config.TestConnectionAsync();
                
                // Health check
                Console.WriteLine("\n📊 Health Check...");
                Console.WriteLine(new string('-', 70));
                var health = await config.HealthCheckAsync();
                var healthJson = JsonConvert.SerializeObject(health, Formatting.Indented);
                Console.WriteLine(healthJson);
                
                Console.WriteLine("\n✅ Configuração concluída!");
            }
            catch (Exception e)
            {
                Console.WriteLine($"\n❌ Erro: {e.Message}");
                Environment.Exit(1);
            }
        }
    }
}

/**
 * .csproj dependencies:
 * 
 * <ItemGroup>
 *   <PackageReference Include="AWSSDK.SecretsManager" Version="3.7.0" />
 *   <PackageReference Include="Npgsql" Version="7.0.0" />
 *   <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
 * </ItemGroup>
 * 
 * Instalação:
 * dotnet add package AWSSDK.SecretsManager
 * dotnet add package Npgsql
 * dotnet add package Newtonsoft.Json
 */

/**
 * Exemplo de uso em ASP.NET Core (Startup.cs ou Program.cs):
 * 
 * public void ConfigureServices(IServiceCollection services)
 * {
 *     var dbConfig = DatabaseConfig.Instance;
 *     var connectionString = dbConfig.GetConnectionStringAsync().Result;
 *     
 *     services.AddDbContext<ApplicationDbContext>(options =>
 *         options.UseNpgsql(connectionString));
 * }
 */

/**
 * Exemplo com Entity Framework Core:
 * 
 * public class ApplicationDbContext : DbContext
 * {
 *     protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
 *     {
 *         var dbConfig = DatabaseConfig.Instance;
 *         var connectionString = dbConfig.GetConnectionStringAsync().Result;
 *         optionsBuilder.UseNpgsql(connectionString);
 *     }
 * }
 */

/**
 * Exemplo com Dapper:
 * 
 * using (var conn = await DatabaseConfig.Instance.GetConnectionAsync())
 * {
 *     var users = await conn.QueryAsync<User>("SELECT * FROM users");
 *     foreach (var user in users)
 *     {
 *         Console.WriteLine(user.Name);
 *     }
 * }
 */

File: DatabaseConfig.php
<?php
/**
 * MaraBet AI - Database Configuration (PHP)
 * Obtém credenciais do RDS PostgreSQL via AWS Secrets Manager
 */

require 'vendor/autoload.php';

use Aws\SecretsManager\SecretsManagerClient;
use Aws\Exception\AwsException;

/**
 * Gerenciador de configuração do banco de dados RDS
 */
class DatabaseConfig
{
    private const SECRET_NAME = 'rds!db-3758a324-12a2-4675-b5ff-b92acdf38483';
    private const REGION = 'eu-west-1';
    private const DATABASE_NAME = 'marabet_production';
    
    private $credentials = null;
    private $pdo = null;
    private static $instance = null;
    
    /**
     * Singleton instance
     */
    public static function getInstance()
    {
        if (self::$instance === null) {
            self::$instance = new self();
        }
        return self::$instance;
    }
    
    private function __construct()
    {
    }
    
    /**
     * Obtém credenciais do AWS Secrets Manager
     * 
     * @return array Credenciais
     * @throws Exception
     */
    public function getSecret()
    {
        $client = new SecretsManagerClient([
            'region' => self::REGION,
            'version' => 'latest'
        ]);
        
        try {
            $result = $client->getSecretValue([
                'SecretId' => self::SECRET_NAME,
            ]);
            
        } catch (AwsException $e) {
            $error = $e->getAwsErrorCode();
            
            if ($error === 'ResourceNotFoundException') {
                throw new Exception("Secret " . self::SECRET_NAME . " não encontrado");
            } elseif ($error === 'InvalidRequestException') {
                throw new Exception("Requisição inválida: " . $e->getMessage());
            } elseif ($error === 'InvalidParameterException') {
                throw new Exception("Parâmetro inválido: " . $e->getMessage());
            } else {
                throw $e;
            }
        }
        
        // Parse JSON
        $secretString = $result['SecretString'];
        $secret = json_decode($secretString, true);
        
        return $secret;
    }
    
    /**
     * Obtém credenciais (com cache)
     */
    public function getCredentials()
    {
        if ($this->credentials === null) {
            $this->credentials = $this->getSecret();
        }
        return $this->credentials;
    }
    
    /**
     * Gera DSN para PostgreSQL (PDO)
     * 
     * @param string|null $database Nome do database
     * @return string DSN
     */
    public function getDsn($database = null)
    {
        $creds = $this->getCredentials();
        $dbName = $database ?? self::DATABASE_NAME;
        
        return sprintf(
            "pgsql:host=%s;port=%s;dbname=%s;sslmode=require",
            $creds['host'],
            $creds['port'],
            $dbName
        );
    }
    
    /**
     * Gera connection string completa
     */
    public function getConnectionString($database = null)
    {
        $creds = $this->getCredentials();
        $dbName = $database ?? self::DATABASE_NAME;
        
        return sprintf(
            "postgresql://%s:%s@%s:%s/%s?sslmode=require",
            $creds['username'],
            $creds['password'],
            $creds['host'],
            $creds['port'],
            $dbName
        );
    }
    
    /**
     * Obtém conexão PDO
     * 
     * @param string|null $database Nome do database
     * @return PDO Conexão PDO
     */
    public function getConnection($database = null)
    {
        $creds = $this->getCredentials();
        $dsn = $this->getDsn($database);
        
        $options = [
            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
            PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
            PDO::ATTR_EMULATE_PREPARES => false,
        ];
        
        return new PDO($dsn, $creds['username'], $creds['password'], $options);
    }
    
    /**
     * Obtém conexão PDO (singleton)
     */
    public function getPdo()
    {
        if ($this->pdo === null) {
            $this->pdo = $this->getConnection();
        }
        return $this->pdo;
    }
    
    /**
     * Testa conexão com o banco
     * 
     * @return bool true se conectou
     */
    public function testConnection()
    {
        try {
            $pdo = $this->getConnection();
            
            // Testar query
            $stmt = $pdo->query('SELECT version()');
            $version = $stmt->fetchColumn();
            
            echo "✅ Conexão bem-sucedida!\n";
            echo "   PostgreSQL: " . substr($version, 0, 50) . "...\n";
            
            return true;
            
        } catch (Exception $e) {
            echo "❌ Erro na conexão: " . $e->getMessage() . "\n";
            return false;
        }
    }
    
    /**
     * Configuração para Laravel (config/database.php)
     */
    public function getLaravelConfig($database = null)
    {
        $creds = $this->getCredentials();
        $dbName = $database ?? self::DATABASE_NAME;
        
        return [
            'driver' => 'pgsql',
            'host' => $creds['host'],
            'port' => $creds['port'],
            'database' => $dbName,
            'username' => $creds['username'],
            'password' => $creds['password'],
            'charset' => 'utf8',
            'prefix' => '',
            'schema' => 'public',
            'sslmode' => 'require',
        ];
    }
    
    /**
     * Imprime informações do banco
     */
    public function printInfo()
    {
        $creds = $this->getCredentials();
        
        echo str_repeat('=', 70) . "\n";
        echo "🗄️  MARABET AI - RDS POSTGRESQL\n";
        echo str_repeat('=', 70) . "\n\n";
        echo "Host:         {$creds['host']}\n";
        echo "Port:         {$creds['port']}\n";
        echo "Username:     {$creds['username']}\n";
        echo "Password:     " . str_repeat('*', strlen($creds['password'])) . "\n";
        echo "Engine:       {$creds['engine']}\n";
        echo "Database:     " . self::DATABASE_NAME . "\n\n";
        echo "Connection String:\n";
        echo str_repeat('-', 70) . "\n";
        echo $this->getConnectionString() . "\n\n";
        echo "DSN (PDO):\n";
        echo str_repeat('-', 70) . "\n";
        echo $this->getDsn() . "\n\n";
        echo str_repeat('=', 70) . "\n";
    }
}

/**
 * Funções de conveniência
 */
function getCredentials()
{
    return DatabaseConfig::getInstance()->getCredentials();
}

function getConnectionString($database = null)
{
    return DatabaseConfig::getInstance()->getConnectionString($database);
}

function getPdo()
{
    return DatabaseConfig::getInstance()->getPdo();
}

function testConnection()
{
    return DatabaseConfig::getInstance()->testConnection();
}

/**
 * Teste do módulo
 */
if (php_sapi_name() === 'cli' && basename(__FILE__) === basename($_SERVER['SCRIPT_FILENAME'])) {
    echo "🚀 MaraBet AI - Database Configuration\n\n";
    
    try {
        $config = DatabaseConfig::getInstance();
        
        // Mostrar informações
        $config->printInfo();
        
        // Testar conexão
        echo "\n🔌 Testando conexão...\n";
        echo str_repeat('-', 70) . "\n";
        $config->testConnection();
        
        echo "\n✅ Configuração concluída!\n";
        
    } catch (Exception $e) {
        echo "\n❌ Erro: " . $e->getMessage() . "\n";
        exit(1);
    }
}

/**
 * composer.json:
 * 
 * {
 *     "require": {
 *         "aws/aws-sdk-php": "^3.0"
 *     }
 * }
 * 
 * Instalação:
 * composer require aws/aws-sdk-php
 */

/**
 * Exemplo de uso em Laravel:
 * 
 * // config/database.php
 * $dbConfig = DatabaseConfig::getInstance();
 * 
 * return [
 *     'connections' => [
 *         'pgsql' => $dbConfig->getLaravelConfig(),
 *     ],
 * ];
 */

/**
 * Exemplo de uso simples:
 * 
 * <?php
 * require 'DatabaseConfig.php';
 * 
 * // Obter PDO
 * $pdo = getPdo();
 * 
 * // Executar query
 * $stmt = $pdo->query('SELECT * FROM users');
 * $users = $stmt->fetchAll();
 * 
 * foreach ($users as $user) {
 *     echo $user['name'] . "\n";
 * }
 */

